hibernate-core/src/main/java/org/hibernate/collection/internal/PersistentBag.java

```diff
@@ -131,21 +134,46 @@ public void beforeInitialize(CollectionPersister persister, int anticipatedSize)
 	}
 
 	@Override
+	@SuppressWarnings( "unchecked" )
 	public boolean equalsSnapshot(CollectionPersister persister) throws HibernateException {
 		final Type elementType = persister.getElementType();
-		final List sn = (List) getSnapshot();
+		final List<Object> sn = (List<Object>) getSnapshot();
 		if ( sn.size() != bag.size() ) {
 			return false;
 		}
-		for ( Object elt : bag ) {
-			final boolean unequal = countOccurrences( elt, bag, elementType ) != countOccurrences( elt, sn, elementType );
-			if ( unequal ) {
+
+		// HHH-11032 - Group objects by Type.getHashCode() to reduce the complexity of the search
+		final Map<Integer, List<Object>> hashToInstancesBag = groupByEqualityHash( bag, elementType );
+		final Map<Integer, List<Object>> HashToInstancesSn = groupByEqualityHash( sn, elementType );
+		if ( hashToInstancesBag.size() != HashToInstancesSn.size() ) {
+			return false;
+		}
+
+		for ( Map.Entry<Integer, List<Object>> hashToInstancesBagEntry : hashToInstancesBag.entrySet() ) {
+			final int hash = hashToInstancesBagEntry.getKey();
```

Comment:
should use `Integer` rather than `int` so to prevent unboxing and then autoboxing again.

Reply:
Thanks, I'll make that change.

Reply:
Yes, I'll make that change.